apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Span.java

```diff
@@ -262,50 +262,66 @@ public void beforeEnd(long epochMicros) {
-                        this.tracer.endSpan(buffered);
+        // Why do we increment references of this span here?
+        // The only thing preventing the "this"-span from being recycled is the initial reference increment in onAfterStart()
+        // There are multiple ways in afterEnd() on how this reference may be decremented and therefore potentially causing recycling:
+        //  - we call tracer.endSpan() for this span and the span is dropped / not reported for some reason
+        //  - we call tracer.endSpan() for this span and the span is reported and released afterwards (=> recycled on the reporter thread!)
+        //  - we successfully set parent.bufferedSpan to "this".
+        //     - a span on a different thread with the same parent can now call tracer.endSpan() for parent.bufferedSpan (=this span)
+        //     - the parent span is ended on a different thread and calls tracer.endSpan() for parent.bufferedSpan (=this span)
+        // By incrementing the reference count here, we guarantee that the "this" span is only recycled AFTER we decrement the reference count again
+        this.incrementReferences();
+        try {
+            if (transaction != null && transaction.isSpanCompressionEnabled() && parent != null) {
+                Span parentBuffered = parent.bufferedSpan.incrementReferencesAndGet();
+                try {
+                    if (parent.isFinished() || !isCompressionEligible()) {
+                        if (parentBuffered != null) {
+                            if (parent.bufferedSpan.compareAndSet(parentBuffered, null)) {
+                                this.tracer.endSpan(parentBuffered);
+                            }
+                        }
+                        this.tracer.endSpan(this);
+                        return;
+                    }
+                    if (parentBuffered == null) {
+                        if (!parent.bufferedSpan.compareAndSet(null, this)) {
+                            // the failed update would ideally lead to a compression attempt with the new buffer,
+                            // but we're dropping the compression attempt to keep things simple
+                            this.tracer.endSpan(this);
+                        }
+                        return;
+                    }
+                    if (!parentBuffered.tryToCompress(this)) {
+                        if (parent.bufferedSpan.compareAndSet(parentBuffered, this)) {
+                            this.tracer.endSpan(parentBuffered);
+                        } else {
+                            // the failed update would ideally lead to a compression attempt with the new buffer,
+                            // but we're dropping the compression attempt to keep things simple
+                            this.tracer.endSpan(this);
+                        }
+                    } else {
+                        if (isSampled() && transaction != null) {
```

Comment:
[For Reviewer]:
Small change of behaviour here. Previously, if the compression was successfull but the span is **not** sampled, we would never
 * decrementReferences() on this
 * decrementReferences() on the parent
 
 I assumed that this was a bug and fixed it here. This behaviour was introduced [here](https://github.com/elastic/apm-agent-java/pull/2477/commits/40b709f5d4ad0fdfd043572ed169be6315946062) and was left uncommented, which is why I assume it was not intentional.´